File: gpio4j/src/main/java/org/overworld/gpio4j/serial/wire/Read.java
package org.overworld.gpio4j.serial.wire;

import org.overworld.gpio4j.serial.LineByte;
import org.overworld.gpio4j.serial.MessageType;

public class Read implements Message {
	
	private Integer pin;

	@Override
	public Message parse(byte[] wireData) {
		
		if (wireData.length != 2) throw new IllegalStateException("Incorrect number of bytes");
		
		pin = new LineByte(wireData[0], wireData[1]).asInt();
		return this;
	}

	@Override
	public byte[] wireData() {
		
		if (pin == null) throw new IllegalStateException("Pin is not set when generating wire data");
		
		LineByte lb = new LineByte(pin);
		return new byte[] { lb.getHi(), lb.getLo() };
	}

	@Override
	public MessageType getType() {
		return MessageType.READ;
	}
}

File: gpioBoardDriver/serial/Read.cpp
#include "Read.h"

class Read {
public:
    Read() : pin_(0) {}
    Read(int pin) : pin_(pin) {}

    Read& parse(const std::vector<uint8_t>& wireData) {
        if (wireData.size() < 2) {
            throw std::runtime_error("Incorrect number of bytes");
        }

        pin_ = LineByte((wireData[0] << 8) | wireData[1]).asInt();
        return *this;
    }

    std::vector<uint8_t> wireData() const {
        if (pin_ == 0) {
            throw std::runtime_error("Pin is not set when generating wire data");
        }

        return LineByte(pin_).getHiLo();
    }

    MessageType getType() const {
        return MessageType::READ;
    }

    int getPin() const {
        return pin_;
    }

    void setPin(int pin) {
        pin_ = pin;
    }

    Read withPin(int pin) const {
        return Read(pin);
    }

private:
    int pin_;
};
